Fix: font-family: initial should apply to the font instead of inherit from ancestor

AlexsanderDamaceno opened the pull request:
#### 0a824d7ee6fb87ea300d96d9f3a9548e44f1bdb0
<pre>
Fix: font-family: initial should apply to the font instead of inherit from ancestor
<a href="https://bugs.webkit.org/show_bug.cgi?id=200709">https://bugs.webkit.org/show_bug.cgi?id=200709</a>

Reviewed by NOBODY (OOPS!).

As &apos;initialDesc.firstFamily().isEmpty()&apos; was returning 0,
we were incorrectly skipping the font family assignment in &apos;setFamilies&apos;,
causing the font family of the previous ancestor to be applied instead.

* LayoutTests/fast/text/font-family-initial-01-expected.txt: Added.
* LayoutTests/fast/text/font-family-initial-01.html: Added.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialFontFamily):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a824d7ee6fb87ea300d96d9f3a9548e44f1bdb0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/173678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/47611 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/41391 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/183252 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/131546 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/175543 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/48903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/47293 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/135055 "Found 2 new test failures: fast/text/font-family-initial-01.html imported/w3c/web-platform-tests/css/css-fonts/inheritance.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/95279 "1 flakes 5 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/176636 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/38483 "Found 1 new test failure: fast/text/font-family-initial-01.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/155839 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/115533 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/37124 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-fonts/inheritance.html (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/35486 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/31736 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/147548 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/35333 "Found 2 new test failures: fast/text/font-family-initial-01.html imported/w3c/web-platform-tests/css/css-fonts/inheritance.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/185678 "Built successfully") | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/38714 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/39686 "Found 2 new test failures: fast/text/font-family-initial-01.html imported/w3c/web-platform-tests/css/css-fonts/inheritance.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/143940 "Found 1 new test failure: fast/text/font-family-initial-01.html (failure)") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/47278 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/155395 "Exiting early after 10 failures. 10 tests run. 1 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/143799 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/47957 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/31949 "Found 2 new test failures: fast/text/font-family-initial-01.html imported/w3c/web-platform-tests/css/css-fonts/inheritance.html (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/113144 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/38721 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/119835 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/46463 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/10288 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/45865 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/46096 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/45924 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ef5e3dba66ebd7ca8257e05144bc7e412065e239. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ef5e3dba66ebd7ca8257e05144bc7e412065e239. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6964b4ec1e74713c756b4784ff2096bf98c81114. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6964b4ec1e74713c756b4784ff2096bf98c81114. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/642d637bbcb62a0e5166098e6d5663bc5c4fcddb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/642d637bbcb62a0e5166098e6d5663bc5c4fcddb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

AlexsanderDamaceno committed:
Fix: font-family: initial should apply to the font instead of inherit from ancestor
https://bugs.webkit.org/show_bug.cgi?id=200709

Reviewed by NOBODY (OOPS!).

As 'initialDesc.firstFamily().isEmpty()' was returning 0,
we were incorrectly skipping the font family assignment in 'setFamilies',
causing the font family of the previous ancestor to be applied instead.

* LayoutTests/fast/text/font-family-initial-01-expected.txt: Added.
* LayoutTests/fast/text/font-family-initial-01.html: Added.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialFontFamily):

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/93cc63987b0c9c73f4a3844239590519270a3446)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78157 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57202 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31533 "Build is in progress. Recent messages:") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80290 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/31533 "Build is in progress. Recent messages:") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/31533 "Build is in progress. Recent messages:") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71299 "Build is in progress. Recent messages:") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/31533 "Build is in progress. Recent messages:") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77393 "Build is in progress. Recent messages:") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/31533 "Build is in progress. Recent messages:") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/55/builds/31533 "Build is in progress. Recent messages:") | | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -743,9 +743,11 @@ inline void BuilderCustom::applyInitialFontFamily(BuilderState& builderState)
             builderState.setFontSize(fontDescription, Style::fontSizeForKeyword(sizeIdentifier, false, builderState.document()));
     }
+
+    auto value = AtomString("initial"_s);
+    initialDesc.setOneFamily(value);
     if (!initialDesc.firstFamily().isEmpty())
         fontDescription.setFamilies(initialDesc.families());
```
```suggestion
    fontDescription.setFamilies(initialDesc.families());
```

This is what I originally suggested. Doesn't this work?

nt1m replied:
I doubt your fix is correct, since it will instead look up a font named "initial".

nt1m replied:
I'm curious what crash you get, because the end result is actually the same, with the only difference that your version sets a bogus font-family before setting the correct one.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d6f39f47e67288a219d73bee05d2c0a87e027c61. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f4e57db1029930e7654c91c25ed839ea4ec75d13. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/aaa1425d682abeee0f7203f24767c99e8d3ffca6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37115

nt1m commented:
I don't think this fix is right, given the test results. The default font should not leak through the computed style.

nt1m commented:
(fwiw, I'm not sure what the right fix is here)

vitorroriz commented:
> @vitorroriz would be ok get initial font from document?

Hi, I'm not sure I understand the question. Could you provide more context?

AlexsanderDamaceno committed:
Fix: font-family: initial should apply to the font instead of inherit from ancestor
https://bugs.webkit.org/show_bug.cgi?id=200709

Reviewed by NOBODY (OOPS!).

As 'initialDesc.firstFamily().isEmpty()' was returning 0,
we were incorrectly skipping the font family assignment in 'setFamilies',
causing the font family of the previous ancestor to be applied instead.

* LayoutTests/fast/text/font-family-initial-01-expected.txt: Added.
* LayoutTests/fast/text/font-family-initial-01.html: Added.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialFontFamily):
